apm-sniffer/apm-sdk-plugin/finagle-6.25.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/finagle/AbstractInterceptor.java

```diff
@@ -0,0 +1,76 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.finagle;
+
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceConstructorInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+
+import java.lang.reflect.Method;
+
+abstract class AbstractInterceptor implements InstanceConstructorInterceptor, InstanceMethodsAroundInterceptor {
+    @Override
+    public void onConstruct(EnhancedInstance objInst, Object[] allArguments) {
+        if (CompatibilityChecker.isCompatible()) {
+            onConstructImpl(objInst, allArguments);
+        }
+    }
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
+        if (CompatibilityChecker.isCompatible()) {
+            beforeMethodImpl(objInst, method, allArguments, argumentsTypes, result);
+        }
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Object ret) throws Throwable {
+        if (CompatibilityChecker.isCompatible()) {
+            return afterMethodImpl(objInst, method, allArguments, argumentsTypes, ret);
+        }
+        return ret;
+    }
+
+    @Override
+    public void handleMethodException(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Throwable t) {
+        if (CompatibilityChecker.isCompatible()) {
+            handleMethodExceptionImpl(objInst, method, allArguments, argumentsTypes, t);
+        }
+    }
+
+    protected void onConstructImpl(EnhancedInstance objInst, Object[] allArguments) {
+
+    }
+
+    protected void beforeMethodImpl(EnhancedInstance objInst, Method method, Object[] allArguments,
+                                    Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
+
+    }
+
+    protected Object afterMethodImpl(EnhancedInstance objInst, Method method, Object[] allArguments,
+                                     Class<?>[] argumentsTypes, Object ret) throws Throwable {
+        return ret;
+    }
+
+    protected void handleMethodExceptionImpl(EnhancedInstance objInst, Method method, Object[] allArguments,
+                                             Class<?>[] argumentsTypes, Throwable t) {
+
+    }
```

Comment:
These methods are overridden by methods in subsequent integrations. Recommend to modify these methods to abstract methods, so that the logic may be clearer.  Or they can be split into two classes?

Reply:
I will make these methods abstract.  

Because `ClientDestTracingFilterInterceptor` need to implement both `onConstructImpl ` and `beforeMethodImpl ` methods and java did not support mutiple inheritance,   so they can not be split into two class.

Reply:
Actually, in java, once you have set `AbstractInterceptor` as abstract, all these empty methods could be deleted directly :)

Reply:
These methods we discussed above are`onConstructImpl`,`beforeMethodImpl `, they are not inherited from interface.  The purpose of  `AbstractInterceptor ` is a adapter for `InstanceConstructorInterceptor , InstanceMethodsAroundInterceptor `.

Reply:
Let's leave this here. My review has been challenging deeper things than this code style  discussion. Please follow those first, then we could do the review again.